### PR TITLE
[Refactor] Unify repo search order by logic

### PR DESCRIPTION
**Pull request by 6543**
have repo OrderBy definitions defined in one place and use a single type for OrderBy database options

### CODE REVIEW

File: models/repo/search.go [hunk 1]
```diff
@@ -5,20 +5,70 @@ package repo
...
+		"id":        SearchOrderByIDReverse,
+		"num_stars": SearchOrderByStarsReverse,
+		"num_forks": SearchOrderByForksReverse,
 	},
 }
+
+// SearchOrderByFlatMap is similar to SearchOrderByMap but use human language keywords
+// to decide between asc and desc
+var SearchOrderByFlatMap = map[string]db.SearchOrderBy{
+	"newest":                SearchOrderByMap["desc"]["created"],
```
Comment (6543) [inline models/repo/search.go:61]: this make it easyer to match human keywords with api options ... but if prefered I could also use the const directly ...

File: models/repo/search.go [hunk 2]
```diff
@@ -5,20 +5,70 @@ package repo
...
-		"size":    db.SearchOrderBySizeReverse,
-		"id":      db.SearchOrderByIDReverse,
+		"alpha":     SearchOrderByAlphabeticallyReverse,
+		"created":   SearchOrderByNewest,
+		"updated":   SearchOrderByRecentUpdated,
+		"size":      SearchOrderBySizeReverse,
+		"git_size":  SearchOrderByGitSizeReverse,
+		"lfs_size":  SearchOrderByLFSSizeReverse,
+		"id":        SearchOrderByIDReverse,
+		"num_stars": SearchOrderByStarsReverse,
```
Comment (6543) [inline models/repo/search.go]: extend options to let api be on pair with webui options

File: models/repo/search.go [hunk 3]
```diff
@@ -5,20 +5,70 @@ package repo
 
 import "code.gitea.io/gitea/models/db"
 
+// Strings for sorting result
+const (
+	SearchOrderByAlphabetically        db.SearchOrderBy = "owner_name ASC, name ASC"
```
Comment (6543) [inline models/repo/search.go:11]: this will affect the explore page similar as  #30784 too

Comment (6543) [reply models/repo/search.go:11]: if desired I'll create a dedicated backport for 1.22 cc @lunny

Comment (lunny) [reply models/repo/search.go:11]: Sorry, I don't think we should backport this to v1.22. Please only backport bugfix since it's a refactor PR.

Comment (6543) [reply models/repo/search.go:11]: @lunny I'll create the pull so you do understand what I ment ...

Comment (6543) [reply models/repo/search.go:11]: -> #30882

**Comment by 6543** [comment]
I would say this refactor is ready ;)

**Comment by lunny** [comment]
> I would say this refactor is ready ;)

I will review and merge after v1.22 stable is released since it's a refactor PR.

**Comment by 6543** [comment]
@lunny we now have v1.22.0 tagged :)

File: models/db/search.go [hunk 4]
```diff
@@ -18,12 +18,6 @@ const (
 	SearchOrderByRecentUpdated         SearchOrderBy = "updated_unix DESC"
 	SearchOrderByOldest                SearchOrderBy = "created_unix ASC"
 	SearchOrderByNewest                SearchOrderBy = "created_unix DESC"
-	SearchOrderBySize                  SearchOrderBy = "size ASC"
```
Comment (lunny) [inline models/db/search.go:21]: Why not keep them here?

Comment (6543) [reply models/db/search.go:21]: because they are not generic but table "repository" specific

Comment (6543) [reply models/db/search.go:21]: of course if there would exist an other table witch we want to order by column 'size' it should belong in generig db search ... but I did not found one - has that changed or is something planed?

**Review by lunny — APPROVED** [review]

File: routers/web/explore/repo.go [hunk 5]
```diff
@@ -61,43 +61,14 @@ func RenderRepoSearch(ctx *context.Context, opts *RepoSearchOptions) {
...
-		orderBy = db.SearchOrderByStarsReverse
-	case "feweststars":
-		orderBy = db.SearchOrderByStars
-	case "mostforks":
-		orderBy = db.SearchOrderByForksReverse
-	case "fewestforks":
-		orderBy = db.SearchOrderByForks
-	default:
-		ctx.Data["SortType"] = "recentupdate"
+	if order, ok := repo_model.SearchOrderByFlatMap[sortOrder]; ok {
```
Comment (delvh) [inline routers/web/explore/repo.go:66]: One potential problem with this approach:
`ASc` or `ASC` results in the default sorting, even if a valid sorting is given
We should probably lowercase the order first

File: models/repo/search.go [hunk 6]
```diff
@@ -5,20 +5,72 @@ package repo
 
 import "code.gitea.io/gitea/models/db"
 
+// Strings for sorting result
+const (
+	// only used for repos
+	SearchOrderByAlphabetically        db.SearchOrderBy = "owner_name ASC, name ASC"
```
Comment (delvh) [inline models/repo/search.go:11]: If it wasn't already present in the original code, I'd propose dropping the `Search-` prefix in this PR.
It only confuses without bringing any additional value.
As it exists already, I'll see that I open another refactoring PR once this one was merged.

**Review by delvh — APPROVED** [review]
(LGTM although we should probably fix the uppercase problem before the merge)

**Comment by 6543** [comment]
@delvh that are two valid refactors :)

But as this is already taking that long i propose to make two small followup pulls adressing it :)

**Comment by delvh** [comment]
The first one is a bug, not a refactor.
And the bug is introduced through this PR.

File: models/repo/search.go (hunk 6 above)
Comment (wxiaoguang) [inline models/repo/search.go:11]: Every `const` here is only used once for constructing the "map"

I think all of these `const` should be removed, and just fill the values into the map directly.

File: models/repo/search.go (hunk 1 above)
Comment (wxiaoguang) [reply models/repo/search.go:61]: > but if prefered I could also use the const directly ...

I also prefer. No need to introduce a one-time const or one-time variable ...........

File: routers/web/explore/repo.go (hunk 5 above)
Comment (6543) [reply routers/web/explore/repo.go:66]: done

File: models/repo/search.go (hunk 6 above)
Comment (6543) [reply models/repo/search.go:11]: well I"ll merge this now ... just create a pull and ping me in it :)

Comment (6543) [reply models/repo/search.go:11]: -> #31359

**Pull request merged**